pkg/gadgets/tcptracer/tracer/core/bpf/tcptracer.bpf.c

```diff
@@ -0,0 +1,359 @@
+{
+	return enter_tcp_connect(ctx, sk);
+}
+
+SEC("kretprobe/tcp_v4_connect")
+int BPF_KRETPROBE(tcp_v4_connect_ret, int ret)
+{
+	return exit_tcp_connect(ctx, ret, AF_INET);
+}
+
+SEC("kprobe/tcp_v6_connect")
+int BPF_KPROBE(tcp_v6_connect, struct sock *sk)
+{
+	return enter_tcp_connect(ctx, sk);
+}
+
+SEC("kretprobe/tcp_v6_connect")
+int BPF_KRETPROBE(tcp_v6_connect_ret, int ret)
+{
+	return exit_tcp_connect(ctx, ret, AF_INET6);
+}
+
+SEC("kprobe/tcp_close")
+int BPF_KPROBE(entry_trace_close, struct sock *sk)
+{
+	__u64 pid_tgid = bpf_get_current_pid_tgid();
+	__u32 pid = pid_tgid >> 32;
+	__u64 uid_gid = bpf_get_current_uid_gid();
+	__u32 uid = uid_gid;
+	struct tuple_key_t tuple = {};
+	struct event event = {};
+	struct task_struct *task;
+	u16 family;
+	u64 mntns_id;
+
+	if (filter_event(sk, uid, pid))
+		return 0;
+
+	/* Don't generate close events for connections that were never
+	 * established in the first place.
+	 */
+	u8 oldstate = BPF_CORE_READ(sk, __sk_common.skc_state);
+	if (oldstate == TCP_SYN_SENT ||
+	    oldstate == TCP_SYN_RECV ||
+	    oldstate == TCP_NEW_SYN_RECV)
+		return 0;
+
+	family = BPF_CORE_READ(sk, __sk_common.skc_family);
+	if (!fill_tuple(&tuple, sk, family))
+		return 0;
+
+	task = (struct task_struct*)bpf_get_current_task();
+	mntns_id = (u64) BPF_CORE_READ(task, nsproxy, mnt_ns, ns.inum);
+
+	fill_event(&tuple, &event, pid, uid, family, TCP_EVENT_TYPE_CLOSE, mntns_id);
+	bpf_get_current_comm(&event.task, sizeof(event.task));
+
+	bpf_perf_event_output(ctx, &events, BPF_F_CURRENT_CPU,
+		      &event, sizeof(event));
+
+	return 0;
+};
+
+SEC("kprobe/tcp_set_state")
+int BPF_KPROBE(enter_tcp_set_state, struct sock *sk, int state)
+{
+	struct tuple_key_t tuple = {};
+	struct event event = {};
+	__u16 family;
+	int i;
+
+	if (state != TCP_ESTABLISHED && state != TCP_CLOSE)
+		goto end;
+
+	family = BPF_CORE_READ(sk, __sk_common.skc_family);
+
+	if (!fill_tuple(&tuple, sk, family))
+		goto end;
+
+	if (state == TCP_CLOSE)
+		goto end;
+
+	struct pid_comm_t *p;
+	p = bpf_map_lookup_elem(&tuplepid, &tuple);
+	if (!p)
+		return 0; /* missed entry */
+
+	fill_event(&tuple, &event, p->pid, p->uid, family, TCP_EVENT_TYPE_CONNECT, p->mntns_id);
+	for (i = 0; i < TASK_COMM_LEN; i++)
+		event.task[i] = p->comm[i];
+
+	bpf_perf_event_output(ctx, &events, BPF_F_CURRENT_CPU,
+		      &event, sizeof(event));
+
+end:
+	bpf_map_delete_elem(&tuplepid, &tuple);
+
+	return 0;
+}
+
+SEC("kretprobe/inet_csk_accept")
+int BPF_KRETPROBE(exit_inet_csk_accept, struct sock *sk)
+{
+	__u64 pid_tgid = bpf_get_current_pid_tgid();
+	__u32 pid = pid_tgid >> 32;
+	__u64 uid_gid = bpf_get_current_uid_gid();
+	__u32 uid = uid_gid;
+	struct task_struct *task;
+	__u16 sport, family;
+	struct event event = {};
+	struct tuple_key_t t = {};
+	u64 mntns_id;
+
+	if (!sk)
+		return 0;
+
+	if (filter_event(sk, uid, pid))
+		return 0;
+
+	family = BPF_CORE_READ(sk, __sk_common.skc_family);
+
+	bool r = fill_tuple(&t, sk, family);
+	if (!r && t.sport != 0)
```

Comment:
![](https://media.makeameme.org/created/when-i-wrote-a49411.jpg)
 
I'll update this because I think it was completely wrong. The goal is to check that IP addresses and ports are not 0 as done in the Python version https://github.com/iovisor/bcc/blob/13b5563c11f7722a61a17c6ca0a1a387d2fa7788/tools/tcptracer.py#L523.